### ISSUE TITLE
Etcd2 -> etcd3 migration is broken

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#36056]
It seems that after etcd2->etcd3 migration all objects are in their initial state (i.e. in the state in which they were created, as if there were no updates at all in the system).

How to repro:
- start a cluster from head
- create some data (e.g. kubectl run pod1 --image=gcr.io/google_containers/pause-amd64:3.0)
- login to master of the cluster
- kill apiserver sudo mv /etc/kubernetes/manifests/kube-apiserver.manifest .
- wait until apiserver is down (so that noone will be writing to etcd)
- open /etc/kubernetes/manifest/etcd.sh and change:
  a) image: to 3.0.13-experimental.2 (from 2.2.1)
  b) TARGET_STORAGE to etcd3
  c) TARGET_VERSION to 3.0.13
- save the manifest and wait until etcd will be restarted and running again (underneath it will perform the migration using
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/migrate-if-needed.sh

@xiang90 @hongchaodeng @mml @lavalamp

**Comment by hongchaodeng** [comment]
Some details are missing. Can you clarify:
- What's the output? What's the expected output? How are they different?
- Can you provide what data you created? Ideally in a script which uses kubectl?

**Comment by wojtek-t** [comment]
1. You can do it by just starting the cluster (e.g. just look into Node objects)
2. the ouput is as I wrote, the object in version from the creation moment (e.g pods are pending, nodes are not ready, without route (ie networkUnavailable=true),etc.)

**Comment by hongchaodeng** [comment]
I'm not getting obvious answers for my questions:

> 1. What's the output? What's the expected output? How are they different?
> 2. Can you provide what data you created? Ideally in a script which uses kubectl?

For (1), can you give some examples?
For (2), can you provide the data and script?

**Comment by wojtek-t** [comment]
Example data:
```
> curl localhost:2379/v2/keys/registry/minions/e2e-test-wojtekt-master
{"action":"get","node":{"key":"/registry/minions/e2e-test-wojtekt-master","value":"{\"kind\":\"Node\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"e2e-test-wojtekt-master\",\"selfLink\":\"/api/v1/nodes/e2e-test-wojtekt-master\",\"uid\":\"ac620924-a0ec-11e6-bef3-42010af00002\",\"creationTimestamp\":\"2016-11-02T11:08:24Z\",\"labels\":{\"beta.kubernetes.io/arch\":\"amd64\",\"beta.kubernetes.io/instance-type\":\"n1-standard-1\",\"beta.kubernetes.io/os\":\"linux\",\"failure-domain.beta.kubernetes.io/region\":\"us-central1\",\"failure-domain.beta.kubernetes.io/zone\":\"us-central1-b\",\"kubernetes.io/hostname\":\"e2e-test-wojtekt-master\"},\"annotations\":{\"volumes.kubernetes.io/controller-managed-attach-detach\":\"true\"}},\"spec\":{\"podCIDR\":\"10.180.2.0/24\",\"externalID\":\"4018458175621854787\",\"providerID\":\"gce://groovy-sentry-504/us-central1-b/e2e-test-wojtekt-master\",\"unschedulable\":true},\"status\":{\"capacity\":{\"alpha.kubernetes.io/nvidia-gpu\":\"0\",\"cpu\":\"1\",\"memory\":\"3788380Ki\",\"pods\":\"110\"},\"allocatable\":{\"alpha.kubernetes.io/nvidia-gpu\":\"0\",\"cpu\":\"1\",\"memory\":\"3788380Ki\",\"pods\":\"110\"},\"conditions\":[{\"type\":\"NetworkUnavailable\",\"status\":\"False\",\"lastHeartbeatTime\":\"2016-11-02T11:09:13Z\",\"lastTransitionTime\":\"2016-11-02T11:09:13Z\",\"reason\":\"RouteCreated\",\"message\":\"RouteController created a route\"},{\"type\":\"OutOfDisk\",\"status\":\"False\",\"lastHeartbeatTime\":\"2016-11-02T11:11:26Z\",\"lastTransitionTime\":\"2016-11-02T11:08:24Z\",\"reason\":\"KubeletHasSufficientDisk\",\"message\":\"kubelet has sufficient disk space available\"},{\"type\":\"MemoryPressure\",\"status\":\"False\",\"lastHeartbeatTime\":\"2016-11-02T11:11:26Z\",\"lastTransitionTime\":\"2016-11-02T11:08:24Z\",\"reason\":\"KubeletHasSufficientMemory\",\"message\":\"kubelet has sufficient memory available\"},{\"type\":\"DiskPressure\",\"status\":\"False\",\"lastHeartbeatTime\":\"2016-11-02T11:11:26Z\",\"lastTransitionTime\":\"2016-11-02T11:08:24Z\",\"reason\":\"KubeletHasNoDiskPressure\",\"message\":\"kubelet has no disk pressure\"},{\"type\":\"InodePressure\",\"status\":\"False\",\"lastHeartbeatTime\":\"2016-11-02T11:11:26Z\",\"lastTransitionTime\":\"2016-11-02T11:08:24Z\",\"reason\":\"KubeletHasNoInodePressure\",\"message\":\"kubelet has no inode pressure\"},{\"type\":\"Ready\",\"status\":\"True\",\"lastHeartbeatTime\":\"2016-11-02T11:11:26Z\",\"lastTransitionTime\":\"2016-11-02T11:08:24Z\",\"reason\":\"KubeletReady\",\"message\":\"kubelet is posting ready status. AppArmor enabled\"}],\"addresses\":[{\"type\":\"InternalIP\",\"address\":\"10.240.0.2\"},{\"type\":\"ExternalIP\",\"address\":\"104.197.109.23\"}],\"daemonEndpoints\":{\"kubeletEndpoint\":{\"Port\":10250}},\"nodeInfo\":{\"machineID\":\"cd99788862101486cf6adab05819c8bd\",\"systemUUID\":\"921C5E2D-3749-8B3E-7794-BACD93DBCA02\",\"bootID\":\"add08964-8eb2-4e15-9f78-7277f1755411\",\"kernelVersion\":\"4.4.21+\",\"osImage\":\"Google Container-VM Image\",\"containerRuntimeVersion\":\"docker://1.11.2\",\"kubeletVersion\":\"v1.5.0-alpha.2.351+07f78836eaf5f6-dirty\",\"kubeProxyVersion\":\"v1.5.0-alpha.2.351+07f78836eaf5f6-dirty\",\"operatingSystem\":\"linux\",\"architecture\":\"amd64\"},\"images\":[{\"names\":[\"gcr.io/google_containers/fluentd-gcp:1.21\"],\"sizeBytes\":498422559},{\"names\":[\"gcr.io/google-containers/kube-addon-manager:v5.2\"],\"sizeBytes\":270266684},{\"names\":[\"gcr.io/google_containers/glbc:0.8.0\"],\"sizeBytes\":243985720},{\"names\":[\"gcr.io/google_containers/kube-apiserver:adca4b14c976e2526b9f2590ccc6863a\"],\"sizeBytes\":122907052},{\"names\":[\"gcr.io/google_containers/rescheduler:v0.2.1\"],\"sizeBytes\":117300467},{\"names\":[\"gcr.io/google_containers/kube-controller-manager:18dd1a82604c3e508680f877b72935c9\"],\"sizeBytes\":101264495},{\"names\":[\"gcr.io/google_containers/kube-scheduler:7f0f253cda8d7658b9ef3839ded02cbe\"],\"sizeBytes\":53405119},{\"names\":[\"gcr.io/google_containers/node-problem-detector:v0.1\"],\"sizeBytes\":40109715},{\"names\":[\"gcr.io/google_containers/etcd:2.2.1\"],\"sizeBytes\":28191895},{\"names\":[\"gcr.io/google_containers/etcd-empty-dir-cleanup:0.0.1\"],\"sizeBytes\":20959104},{\"names\":[\"gcr.io/google_containers/pause-amd64:3.0\"],\"sizeBytes\":746888}]}}\n","modifiedIndex":498,"createdIndex":27}}


> docker exec -t -i <etcd container> /bin/sh
> ETCDCTL_API=3 /usr/local/bin/etcdctl-3.0.13 get /registry/minions/e2e-test-wojtekt-master /registry/minions/e2e-test-wojtekt-master2
/registry/minions/e2e-test-wojtekt-master
{"kind":"Node","apiVersion":"v1","metadata":{"name":"e2e-test-wojtekt-master","uid":"ac620924-a0ec-11e6-bef3-42010af00002","creationTimestamp":"2016-11-02T11:08:24Z","labels":{"beta.kubernetes.io/arch":"amd64","beta.kubernetes.io/instance-type":"n1-standard-1","beta.kubernetes.io/os":"linux","failure-domain.beta.kubernetes.io/region":"us-central1","failure-domain.beta.kubernetes.io/zone":"us-central1-b","kubernetes.io/hostname":"e2e-test-wojtekt-master"},"annotations":{"volumes.kubernetes.io/controller-managed-attach-detach":"true"}},"spec":{"externalID":"4018458175621854787","providerID":"gce://groovy-sentry-504/us-central1-b/e2e-test-wojtekt-master","unschedulable":true},"status":{"capacity":{"alpha.kubernetes.io/nvidia-gpu":"0","cpu":"1","memory":"3788380Ki","pods":"110"},"allocatable":{"alpha.kubernetes.io/nvidia-gpu":"0","cpu":"1","memory":"3788380Ki","pods":"110"},"conditions":[{"type":"NetworkUnavailable","status":"True","lastHeartbeatTime":null,"lastTransitionTime":"2016-11-02T11:08:23Z","reason":"NoRouteCreated","message":"Node created without a route"},{"type":"OutOfDisk","status":"False","lastHeartbeatTime":"2016-11-02T11:08:24Z","lastTransitionTime":"2016-11-02T11:08:24Z","reason":"KubeletHasSufficientDisk","message":"kubelet has sufficient disk space available"},{"type":"MemoryPressure","status":"False","lastHeartbeatTime":"2016-11-02T11:08:24Z","lastTransitionTime":"2016-11-02T11:08:24Z","reason":"KubeletHasSufficientMemory","message":"kubelet has sufficient memory available"},{"type":"DiskPressure","status":"False","lastHeartbeatTime":"2016-11-02T11:08:24Z","lastTransitionTime":"2016-11-02T11:08:24Z","reason":"KubeletHasNoDiskPressure","message":"kubelet has no disk pressure"},{"type":"InodePressure","status":"False","lastHeartbeatTime":"2016-11-02T11:08:24Z","lastTransitionTime":"2016-11-02T11:08:24Z","reason":"KubeletHasNoInodePressure","message":"kubelet has no inode pressure"},{"type":"Ready","status":"True","lastHeartbeatTime":"2016-11-02T11:08:24Z","lastTransitionTime":"2016-11-02T11:08:24Z","reason":"KubeletReady","message":"kubelet is posting ready status. AppArmor enabled"}],"addresses":[{"type":"InternalIP","address":"10.240.0.2"},{"type":"ExternalIP","address":"104.197.109.23"}],"daemonEndpoints":{"kubeletEndpoint":{"Port":10250}},"nodeInfo":{"machineID":"cd99788862101486cf6adab05819c8bd","systemUUID":"921C5E2D-3749-8B3E-7794-BACD93DBCA02","bootID":"add08964-8eb2-4e15-9f78-7277f1755411","kernelVersion":"4.4.21+","osImage":"Google Container-VM Image","containerRuntimeVersion":"docker://1.11.2","kubeletVersion":"v1.5.0-alpha.2.351+07f78836eaf5f6-dirty","kubeProxyVersion":"v1.5.0-alpha.2.351+07f78836eaf5f6-dirty","operatingSystem":"linux","architecture":"amd64"},"images":[{"names":["gcr.io/google-containers/kube-addon-manager:v5.2"],"sizeBytes":270266684},{"names":["gcr.io/google_containers/kube-apiserver:adca4b14c976e2526b9f2590ccc6863a"],"sizeBytes":122907052},{"names":["gcr.io/google_containers/kube-controller-manager:18dd1a82604c3e508680f877b72935c9"],"sizeBytes":101264495},{"names":["gcr.io/google_containers/kube-scheduler:7f0f253cda8d7658b9ef3839ded02cbe"],"sizeBytes":53405119},{"names":["gcr.io/google_containers/etcd:2.2.1"],"sizeBytes":28191895},{"names":["gcr.io/google_containers/etcd-empty-dir-cleanup:0.0.1"],"sizeBytes":20959104},{"names":["gcr.io/google_containers/pause-amd64:3.0"],"sizeBytes":746888}]}}
```

(2) - I don't understand the question - I gave exact repro case in the first post.

**Comment by wojtek-t** [comment]
Talked with @xiang90 offline and he suggested that there is a bug in ```etcdctl migrate``` that doesn't work if there is no snapshot. He suggested forcing snapshot by setting:
```--snapshot-count=1```
and starting etcd to force snapshot.

I did some quick experiment and indeed it seems to help.

I will try to fix this on our end (at least temporarily, until the fix is done in etcdctl).

**Comment by wojtek-t** [comment]
Actually, let me check if this isn't because of simulatnous migration of etcd and even-etcd.

**Comment by wojtek-t** [comment]
Unfortunately it's not related to that. It break even if I'm migrading only one etcd instance.

**Comment by mml** [comment]
@wojtek-t does it reliably break every time, or just sometimes?

**Comment by xiang90** [comment]
@mml It breaks when we migrate newly created clusters. etcd migration tool does not handle pure WAL migration (no checkpointing has been made) well enough.

**Comment by wojtek-t** [comment]
@xiang90 - but it seems it also has some problems with empty(?) wal file (i.e. right after snapshotting)?

**Comment by wojtek-t** [comment]
[The second problem I pasted above with the workaround of --snapshot-count=1 I was trying to make]

**Comment by wojtek-t** [comment]
But still - both these seems like kind of corner-cases...

**Comment by mml** [comment]
The second case seems more likely, because it occurs once every time we make a snapshot, right?

**Comment by xiang90** [comment]
@mml We will fix both at etcd side.

**Comment by wojtek-t** [comment]
@xiang90 has a PR in etcd that seems to fix the problems: https://github.com/coreos/etcd/pull/6794
I built the custom etcdctl version with the patch and it seems to help.

Until etcd release is not done, I've build "3.0.13-experimental.3" image with the custom etcdctl, to allow further testing.

**Comment by hongchaodeng** [comment]
Isn't this fixed?

**Comment by wojtek-t** [comment]
This one is fixed in 3.0.14 (which was release on Friday evening).

**Comment by wojtek-t** [comment]
OK - I've built 3.0.14-experimental.1 image, which our tests are using now. So closing this one (at least for now).
